.github/workflows/provenance.yml

```diff
@@ -0,0 +1,152 @@
+        uses: actions/checkout@ec3a7ce113134d7a93b817d10a8272cb61118579 # v2.3.4
+        with:
+          fetch-depth: 0
+          repository: "${{ env.GENERATOR_REPOSITORY }}"
+          ref: "${{ env.GENERATOR_REF }}"
+
+      - name: Set up golang environment
+        uses: actions/setup-go@424fc82d43fa5a37540bae62709ddcc23d9520d4 # v2.1.3
+        with:
+          go-version: 1.18
+
+      - name: Download dependencies
+        shell: bash
+        run: |
+          set -euo pipefail
+          #TODO(reproducible)
+          go mod vendor
+
+      # TODO(hermeticity) OS-level.
+
+      - name: Build slsa-github-generator
+        shell: bash
+        id: generator-gen
+        run: |
+          set -euo pipefail
+
+          # https://go.dev/ref/mod#build-commands.
+          go build -mod=vendor -o slsa-github-generator github.com/slsa-framework/slsa-github-generator/cmd/slsa-github-generator
+          GENERATOR_DIGEST=$(sha256sum slsa-github-generator | awk '{print $1}')
+          echo "::set-output name=go-generator-sha256::$GENERATOR_DIGEST"
+          echo "hash of slsa-github-generator is $GENERATOR_DIGEST"
+
+      - name: Upload slsa-github-generator
+        uses: actions/upload-artifact@82c141cc518b40d92cc801eee768e7aafc9c2fa2 # v2.3.1
+        with:
+          name: "slsa-github-generator"
+          path: "slsa-github-generator"
+          if-no-files-found: error
+          retention-days: 5
+
+  ###################################################################
+  #                                                                 #
+  #                 Generate the SLSA provenance                    #
+  #                                                                 #
+  ###################################################################
+  provenance:
+    runs-on: ubuntu-latest
+    needs: [generator]
+    permissions:
+      id-token: write # Needed for keyless.
+      contents: read
+    steps:
+      - name: Download generator
+        uses: actions/download-artifact@f023be2c48cc18debc3bacd34cb396e0295e2869 # v2.1.0
+        with:
+          name: "slsa-github-generator"
+
+      - name: Verify slsa-github-generator
+        env:
+          GENERATOR_HASH: "${{ needs.generator.outputs.go-generator-sha256 }}"
+        run: |
+          set -euo pipefail
+
+          echo "hash of slsa-github-generator should be $GENERATOR_HASH"
+          COMPUTED_HASH=$(sha256sum slsa-github-generator | awk '{print $1}')
+          echo "slsa-github-generator hash computed is $COMPUTED_HASH"
+
+          # Compare hashes. Explicit exit to be safe.
+          # TODO: remove the generator as well.
+          echo "$GENERATOR_HASH slsa-github-generator" | sha256sum --strict --check --status || exit -2
+
+          # Make the binary executable.
+          chmod a+x slsa-github-generator
+
+      - name: Create and sign provenance
+        id: sign-prov
+        shell: bash
+        env:
+          SUBJECTS: "${{ inputs.artifacts }}"
```

Comment:
Not needed? isn't this already passed in via `--subjects`?

Reply:
the input to the `--subject` is set by this `$SUBJECT` env variable. Maybe it's misleading, but the code does:
1. Set a `SUBJECTS` env variable for the script, with the value being the input `artifacts` passed to the reusable workflow
2. Run `... -- subjects $SUBJECT` to pass it as argument to the CLI

Does this clarify?

Reply:
Oh, I see. But why not use `${{ toJSON(github) }}` directly in the command? Is it due to shell quoting, such that using `env` avoids the issues? If so, that would be good to comment since it's not obvious to me as a reader.

Reply:
it's good practice to set these via env variables to avoid script injection. The way these scrips work is that GitHub replaces the content with the variable and create the script on disk with the result. So even escaping the variable as `"$UNTRUSTED_VARIABLE"` does not prevent injection. The only way to escape is to:
1. Use the env variable
2. Use `"$VAR"` as well

See https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#understanding-the-risk-of-script-injections for more details. Fyi, scorecard has a check called [`Danegrous`](https://github.com/ossf/scorecard/blob/main/docs/checks.md#dangerous-workflow) to catch these too.

Let's maybe add a comment and a link to the GitHub doc?